File: C++/Data Structures/ArrayStack.cpp
#include <iostream>
#include "abstract.hpp"

using namespace std;

template <class T> class ArrayStack : public Stack<T>
{
private:
    int top, scope;
    T* stack;

public:

    //Contructor
    ArrayStack( int initialCapacity = 8 )
    {
        if( initialCapacity < 0 )
        {
            cerr << "The initial capacity is invalid:" << initialCapacity;
            throw new out_of_range( "" );
        }
        scope = initialCapacity;
        stack = new T[ initialCapacity ];
        top = -1;
    }

    // Methods
    bool isEmpty()
    {
        return top == -1;
    }

    T peek()
    {
        if( isEmpty() )
        {
            cerr << "The stack is Empty" << endl;
            throw new out_of_range( "" );
        }
        return stack[top];
    }

    void push( T theElement )
    {
        if( top == scope - 1 )
        {
            T *old = stack;
            stack = new T[ 2 * scope ];
            for( int i = 0; i < scope; i++ )
                stack[ i ] = old[ i ];
            scope *= 2;
        }
        stack[ ++top ] = theElement;
    }
    T pop()
    {
        if( isEmpty() )
        {
            cerr << "The stack is Empty" << endl;
            throw new out_of_range( "" );
        }
        T topElement = stack[ top-- ];
        return topElement;
    }
};


int main()
{
    cout << boolalpha << endl;
    ArrayStack<int> *numbers = new ArrayStack<int>();
    cout << "is Empty? " << numbers->isEmpty() << endl;
    cout << "15 elements are inserted from 0 ..." << endl;
    int i;
    for( i = 0; i < 15; i++ )
        numbers->push( i );
    cout << "The element in top is: " << numbers->peek() << endl;
    cout << "5 elements are removed" << endl;
    i = 5;
    while( i-- )
        numbers->pop();
    cout << "5 elements are inserted from 15" << endl;
    for( i = 15; i < 20; i++ )
        numbers->push( i );
    cout << "All elements are removed ... " << endl;
    while( !numbers->isEmpty() )
        cout << numbers->pop() << " ";
    cout << endl;
    cout << noboolalpha << endl;
    delete numbers;
    return 0;
}

File: Java/Course Data Structures/Java/unal/datastructures/ArrayStack.java
/** a stack class that uses a one-dimensional array */

package unal.datastructures;

import java.util.*;

public class ArrayStack<T> implements Stack<T>
{
   // fields
   int top;    // current top of stack
   T[] stack;  // element array

   // constructors
   /** create a stack with the given initial capacity
    * @throws IllegalArgumentException when initialCapacity < 1 */
   @SuppressWarnings( "unchecked" )
   public ArrayStack( int initialCapacity )
   {
      if( initialCapacity < 1 )
         throw new IllegalArgumentException
            ( "initialCapacity must be >= 1" );
      stack = ( T[] ) new Object[ initialCapacity ];
      top = -1;
   }

   /** create a stack with initial capacity 10 */
   public ArrayStack( )
   {
      this( 10 );
   }

   // methods
   /** @return true iff stack is empty */
   public boolean isEmpty( )
   {
      return top == -1;
   }

   /** @return top element of stack
    * @throws EmptyStackException when the stack is empty */
   public T peek( )
   {
      if( isEmpty( ) )
         throw new EmptyStackException( );
      return stack[ top ];
   }

   /** add theElement to the top of the stack */
   @SuppressWarnings( "unchecked" )
   public void push( T theElement )
   {
      // increase array size if necessary
      if( top == stack.length - 1 )
      {
         T[] old = stack;
         stack = ( T[] ) new Object[ 2 * stack.length ];
         System.arraycopy( old, 0, stack, 0, old.length );
      }

      // put theElement at the top of the stack
      stack[ ++top ] = theElement;
   }

   /** remove top element of stack and return it
    * @throws EmptyStackException when the stack is empty */
   public T pop( )
   {
      if( isEmpty( ) )
         throw new EmptyStackException( );
      T topElement = stack[ top ];
      stack[ top-- ] = null; // enable garbage collection
      return topElement;
   }

   /** test program */
   public static void main( String[] args )
   {
      int x;
      ArrayStack<Integer> s = new ArrayStack<>( 3 );

      // add a few elements
      s.push( new Integer( 1 ) );
      s.push( new Integer( 2 ) );
      s.push( new Integer( 3 ) );
      s.push( new Integer( 4 ) );

      // delete all elements
      while( !s.isEmpty( ) )
      {
         System.out.println( "Top element is " + s.peek( ) );
         System.out.println( "Removed the element " + s.pop( ) );
      }
   }
}
